protocol.nls

```diff
@@ -381,4 +384,53 @@ to heatmap [including-attacks?]
     ]
     set pcolor scale-color my-color realized-score 0 max-score
   ]
-end
\ No newline at end of file
+end
+
+
+
+
+
+; knowledge and belief tracking during the run . Called during 
+; `set-current-theory-info`. Arguments: 
+; #theory: which theory is currently processed (integer)
+; #non-admiss-argu: number of non admissible arguments for the current theory 
+; (integer)
+; curtheory: the (start x) for the current theory (turtle)
+to track-knowledge [#theory #non-admiss-argu curtheory]
+  let cur-group-id group-id
+  let known-args subjective-arguments with [mytheory = curtheory]
+  let #known-args count known-args
+  ; division by 10 to keep numbers compact as colors are multiple of 10
+  let knowledge sum [(cyan - item cur-group-id group-color-mem) / 10] of 
+    known-args
+  let #defargs #known-args - #non-admiss-argu
+  ; the +1 for #theory is for readability: so the first theory is named "1"
+  let new-knowledge-entry (list (ticks + 1) cur-group-id (#theory + 1) #defargs 
+    #known-args knowledge)
+  set g-knowledge lput new-knowledge-entry g-knowledge
+end
+
+
+
+
+
+; saves the information tracked by `track-knowledge` during the run to a csv 
+; file at the end of the run
+to save-tracked-knowledge
+  let sorted-admissibility map [start-x -> [objective-admissibility] of start-x]
+    sort starts
+  let random-start one-of starts
+  let #args-per-th count startsargum with [mytheory = (random-start)]
+  let filename behaviorspace-experiment-name
+  set filename (word filename "-in-run-knowledge.csv")
+  file-open filename
```

Comment:
This part is currently not multi threading save it seems: When multiple threads try to write to the same file there is no mechanism to ensure proper fencing between those writes and as a result the output can get corrupted